numpy/ma/core.py

```diff
@@ -6779,18 +6778,19 @@ def where(condition, x=_NoValue, y=_NoValue):
         ndtype = np.find_common_type([xv.dtype, yv.dtype], [])
 
     # Construct an empty array and fill it
-    d = np.empty(fc.shape, dtype=ndtype).view(MaskedArray)
-    np.copyto(d._data, xv.astype(ndtype), where=fc)
-    np.copyto(d._data, yv.astype(ndtype), where=notfc)
+    data = np.where(fc, xv.astype(ndtype), yv.astype(ndtype))
+    d = data.view(MaskedArray)
 
     # Create an empty mask and fill it
-    mask = np.zeros(fc.shape, dtype=MaskType)
-    np.copyto(mask, getmask(x), where=fc)
-    np.copyto(mask, getmask(y), where=notfc)
-    mask |= getmaskarray(condition)
+    mask = np.where(fc, getmaskarray(x), getmaskarray(y))
```

Comment:
Sorry for not simply trying, but a bit in a rush: would you not need to create a new mask type, in analogy with the `ndtype` above? I.e., will this work as expected when one or both of the mask arrays have a record dtype?